### PR TITLE
util/log: fix windows color detection crash

### PR DESCRIPTION
**Pull request by BramGruneir**
Until go1.9 this bug https://github.com/golang/go/issues/14853 needs to be handled.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/14483)

### CODE REVIEW

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/util/log/clog.go, line 798 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483#-KgVsinBqfpRqmY68tho:-KgVsinBqfpRqmY68thp:b-tdnyqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cecce548eb0457fc139a5855d83934bc9231d03/pkg/util/log/clog.go#L798)):*
> ```Go
> 			if err != nil {
> 				// Stat() will return an error on Windows in both Powershell and
> 				// console until go1.9. See https://github.com/golang/go/issues/14853
> ```

add a period, a blank comment line (to preserve the paragraph), and a TODO (separated by another blank line).

---

*[pkg/util/log/clog.go, line 804 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483#-KgVsvTsdqQNoDTnnh7C:-KgVsvTsdqQNoDTnnh7D:b2hp06q) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cecce548eb0457fc139a5855d83934bc9231d03/pkg/util/log/clog.go#L804)):*
> ```Go
> 					// Powershell supports colorProfile256. Sadly, detecting the
> 					// shell is not well supported. If there is a requirement to
> 					// run this in a cmd console, the --no-color flag should be
> ```

I think we should be conservative and default to no color on windows.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483#-:-KgVt2lw0olV5bLbk0Nb:bnfp4nl)*

**Comment by BramGruneir** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/util/log/clog.go, line 798 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483#-KgVsinBqfpRqmY68tho:-KgVv-zhb45KBMJlC3HM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cecce548eb0457fc139a5855d83934bc9231d03/pkg/util/log/clog.go#L798)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

add a period, a blank comment line (to preserve the paragraph), and a TODO (separated by another blank line).
</blockquote></details>

Done.

---

*[pkg/util/log/clog.go, line 804 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483#-KgVsvTsdqQNoDTnnh7C:-KgVtwxxoKC1uGHFuzfg:be9jce8) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cecce548eb0457fc139a5855d83934bc9231d03/pkg/util/log/clog.go#L804)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think we should be conservative and default to no color on windows.
</blockquote></details>

Since I'm the one debugging and in our docs we already recommend using PowerShell, I think we're safe making this the default.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/util/log/clog.go, line 804 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483#-KgVsvTsdqQNoDTnnh7C:-KgW0QjEmM1Ltm2o4tPU:b-zdqgrs) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cecce548eb0457fc139a5855d83934bc9231d03/pkg/util/log/clog.go#L804)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Since I'm the one debugging and in our docs we already recommend using PowerShell, I think we're safe making this the default.
</blockquote></details>

Let me rephrase: how does console's lack of color support manifest?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483)*

**Comment by bdarnell** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/util/log/clog.go, line 804 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483#-KgVsvTsdqQNoDTnnh7C:-KgW0s2lq-o8i7CFw2Xe:b-ihzsnt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cecce548eb0457fc139a5855d83934bc9231d03/pkg/util/log/clog.go#L804)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Let me rephrase: how does console's lack of color support manifest?
</blockquote></details>

If this manifests as visible garbage in the terminal, that sounds like a bad experience and it would be better to default to no color. I've also heard via https://github.com/tornadoweb/tornado/pull/1817 that there are differences in color terminal support in different windows versions. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483)*

**Comment by BramGruneir** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/util/log/clog.go, line 804 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483#-KgVsvTsdqQNoDTnnh7C:-Kgyitqr6SYjwpggkqTY:b-mlt1d5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cecce548eb0457fc139a5855d83934bc9231d03/pkg/util/log/clog.go#L804)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

If this manifests as visible garbage in the terminal, that sounds like a bad experience and it would be better to default to no color. I've also heard via https://github.com/tornadoweb/tornado/pull/1817 that there are differences in color terminal support in different windows versions. 
</blockquote></details>

I've changed the default to no colour for windows, but I don't think  we need to support anything besides powershell at this point.  And I don't see any reason to need to support anything besides windows 10 and server 2016, which should have the same colour schemes.

I tried out some of the go libraries that supposedly detected colours available for console and while they claimed windows support I couldn't get them to work.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/util/log/clog.go, line 804 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483#-KgVsvTsdqQNoDTnnh7C:-KgykJoky5wGD5wROLCY:boihlvg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cecce548eb0457fc139a5855d83934bc9231d03/pkg/util/log/clog.go#L804)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I've changed the default to no colour for windows, but I don't think  we need to support anything besides powershell at this point.  And I don't see any reason to need to support anything besides windows 10 and server 2016, which should have the same colour schemes.

I tried out some of the go libraries that supposedly detected colours available for console and while they claimed windows support I couldn't get them to work.
</blockquote></details>

Can you list some of those libraries? Also, we shouldn't just be ignoring the error here - perhaps on Windows it's fine, but elsewhere we should either panic (which is what happens today, albiet without a useful message), or log the error.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483)*

**Comment by BramGruneir** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/util/log/clog.go, line 804 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483#-KgVsvTsdqQNoDTnnh7C:-Kgyu0oGuBiwZ5ZK4_RX:b9l4qa4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cecce548eb0457fc139a5855d83934bc9231d03/pkg/util/log/clog.go#L804)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Can you list some of those libraries? Also, we shouldn't just be ignoring the error here - perhaps on Windows it's fine, but elsewhere we should either panic (which is what happens today, albiet without a useful message), or log the error.
</blockquote></details>

Sure, I took a look at https://github.com/fatih/color which uses https://github.com/mattn/go-isatty.  Just trying to see if it was able to detect PS vs Console vs MSIS/Cygwin.  And it had issues with PS and Console.  Didn't spend too long, as it was clearly a rabbit hole.  It's possible to display colours in Console but just not using the codes we're using.

I've added a panic if not windows.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14483)*
